File: python/tests/test_trie.py
import os
import sys
import unittest
sys.path.append(os.path.join(os.getcwd(), os.path.pardir))

from data_structures.trie import Trie

class TrieTest(unittest.TestCase):

    def test_insert_find(self):
        print("Testing insert/find")
        t = Trie()

        t["bar"] = 1
        t["baz"] = 2
        t["foo"] = 0

        self.assertEqual(1, t.get("bar"))
        self.assertEqual(0, t.get("foo"))
        self.assertEqual(2, t.get("baz"))

        self.assertEqual(1, t["bar"])
        self.assertEqual(0, t["foo"])
        self.assertEqual(2, t["baz"])

    def test_has_prefix(self):
        print("Testing has prefix")
        t = Trie()

        t["bar"] = 1
        t["baz"] = 2
        t["foo"] = 0

        self.assertEqual(True, t.has_prefix("b"))
        self.assertEqual(True, t.has_prefix("ba"))
        self.assertEqual(True, t.has_prefix("bar"))
        self.assertEqual(True, t.has_prefix("baz"))
        self.assertEqual(True, t.has_prefix("f"))
        self.assertEqual(False, t.has_prefix("bat"))
        self.assertEqual(False, t.has_prefix("a"))


if __name__ == '__main__':
    unittest.main()

File: cpp/tests/test_trie.cpp
#include "catch.hpp"
#include "../data_structures/trie/Trie.cpp"
#include <string>

TEST_CASE("Test insert, contains", "[Trie]" ) {
    Trie<int> trie;
    
    trie.insert("f", 0);
    trie.insert("fo", 1);
    trie.insert("foo", 3);
    trie.insert("foob", 4);
    trie.insert("fooba", 5);
    trie.insert("foobar", 6);
    trie.insert("b", 7);
    trie.insert("ba", 8);
    trie.insert("baz", 9);

    REQUIRE(trie.contains("f"));
    REQUIRE(trie.contains("fo"));
    REQUIRE(trie.contains("foo"));
    REQUIRE(trie.contains("foob"));
    REQUIRE(trie.contains("fooba"));
    REQUIRE(trie.contains("foobar"));
    REQUIRE(trie.contains("b"));
    REQUIRE(trie.contains("ba"));
    REQUIRE(trie.contains("baz"));
}

TEST_CASE("Test erase", "[Trie]" ) {
    Trie<int> trie;
    
    trie.insert("f", 0);
    trie.insert("fo", 1);
    trie.insert("foo", 3);
    trie.insert("foob", 4);
    trie.insert("fooba", 5);
    trie.insert("foobar", 6);
    trie.insert("b", 7);
    trie.insert("ba", 8);
    trie.insert("baz", 9);
    
    REQUIRE(trie.contains("f"));
    trie.erase("f");
    REQUIRE(!trie.contains("f"));

    REQUIRE(trie.contains("fo"));
    trie.erase("fo");
    REQUIRE(!trie.contains("fo"));

    REQUIRE(trie.contains("foobar"));
    trie.erase("foobar");
    REQUIRE(!trie.contains("foobar"));

    REQUIRE(trie.contains("fooba"));
    trie.erase("fooba");
    REQUIRE(!trie.contains("fooba"));

    REQUIRE(trie.contains("foo"));
    trie.erase("foo");
    REQUIRE(!trie.contains("foo"));

    REQUIRE(trie.contains("foob"));
    trie.erase("foob");
    REQUIRE(!trie.contains("foob"));

    REQUIRE(trie.contains("b"));
    trie.erase("b");
    REQUIRE(!trie.contains("b"));

    REQUIRE(trie.contains("ba"));
    trie.erase("ba");
    REQUIRE(!trie.contains("ba"));
    
    REQUIRE(trie.contains("baz"));
    trie.erase("baz");
    REQUIRE(!trie.contains("baz"));
}
